test/server.go

```diff
@@ -21,13 +19,14 @@ func CoreDNSServer(corefile string) (*caddy.Instance, error) {
 	defer mu.Unlock()
 	caddy.Quiet = true
 	dnsserver.Quiet = true
-	log.SetOutput(ioutil.Discard)
-
 	return caddy.Start(NewInput(corefile))
 }
 
 // CoreDNSServerStop stops a server.
-func CoreDNSServerStop(i *caddy.Instance) { i.Stop() }
+func CoreDNSServerStop(i *caddy.Instance) {
+	i.Stop()
+	i.ShutdownCallbacks()
```

Comment:
Please remove this. 

These instance starting never exectuted any of these (and some tests work around that), should just add this with this PR

Reply:
CoreDNSServerStop was not used so far ...
and if you do not call ShutdownCallback then the onFinalShutdown functions will not be called : the stop is not complete.